@here/harp-mapview/lib/ClipPlanesEvaluator.ts

```diff
@@ -0,0 +1,524 @@
+
+        let nearPlane: number = this.nearMin;
+        let farPlane: number = this.nearMin + this.planesMaxDistance;
+
+        // This solution computes near and far plane for a set up where
+        // the camera is looking at the center of the scene.
+
+        //         , - ~ ~ ~ - ,
+        //     , '               ' ,
+        //   ,                       ,
+        //  ,                         ,
+        // ,                           ,
+        // ,             . O           ,
+        // ,             | .           ,
+        //  ,            |   .  r     ,
+        //   ,           |      .    ,
+        //     ,         |        , '_____ far
+        //       ' -_, _ | _ ,  ' / T
+        //     near      |      /
+        //               |    / t
+        //             d |  /
+        //               |/
+        //               C
+
+        // The near plane calculus is quite straight forward and works the same as for planar
+        // projections. We simply search for the closest point of the ground just above
+        // the camera, then we apply margin to it along the sphere surface normal:
+        nearPlane = this.getCameraAltitude(camera, projection) - this.nearMargin;
+
+        // The far plane distace calculus requires finding the sphere tangent line that is
+        // co-linear with (goes thru) camera position, such tangent creates right angle
+        // with sphere diameter where it touches its surface (point T). Given that sphere is
+        // always at world orgin and camera orbits around it we have:
+        // angle(OTC) = 90
+        // sin(OCT) = sin(alpha) = r / d
+        // alpha = asin(r / d)
+        const r = EarthConstants.EQUATORIAL_RADIUS;
+        const d = camera.position.length();
+        const alpha = Math.asin(r / d);
+        // The distance to tangent point may be descibed as:
+        // t = sqrt(d^2 - r^2)
+        const t = Math.sqrt(d * d - r * r);
+        // Next step is to project CT vector onto camera eye (forward) vector to get maximum
+        // camera far plane distance. Point on sphere beyond that point won't be visible anyway
+        // unless they are above the ground surface. For such cases [[this.farMargin]] applies.
+
+        // Extract camera X, Y, Z orientation axes into tmp vectors array.
+        camera.matrixWorld.extractBasis(
+            this.m_tmpVectors[0],
+            this.m_tmpVectors[1],
+            this.m_tmpVectors[2]
+        );
+        // Setup quaternion (based on X axis) for angle between tangent and camera eye.
+        this.m_tmpQuaternion.setFromAxisAngle(this.m_tmpVectors[0], alpha);
+        // Acquire camera (eye) forward vector from Z axis reversed (keep it in tmpVectors[2]).
+        const cameraFwdVec = this.m_tmpVectors[2].negate();
+        // Apply quaternion to forward vector, creating tangent vector (store in tmpVectors[1]).
+        const tangentVec = this.m_tmpVectors[1]
```

Comment:
Just keep in mind that it is a matter of representative graph (explanation). I've gone to the same assumption, but working with vectors whenever possible seems to be more self explanatory and error pron (trigonometry constraints). At the end:
farPlane = tangentVec.dot(cameraFwdVec),
it is nothing else then:
farPlane = cos(alpha) * lenght(tangentVec), because here cameraFwdVec is unit length.

Reply:
Agreed, the dot product is easy to understand in this case.